### ISSUE TITLE
Inquiry about RT-DETR Inference Output Limitation (300 Objects per Image)

### ISSUE THREAD

**Issue by HD-L** [ultralytics/ultralytics#20688]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I hope this message finds you well. I am currently working with the RT-DETR model (specifically the RTDETR-l architecture) from the Ultralytics repository and have encountered a question regarding the inference output behavior.

​Issue Description:
During both training and inference, the model consistently returns ​a maximum of 300 predictions per image, even in scenarios where the input image contains a higher number of potential objects. From the documentation and prior discussions, I understand this is related to the num_queries parameter in the RTDETRDecoder module.

​Steps I’ve Taken:

Modified the model configuration (rtdetr.yaml) by explicitly setting num_queries in the decoder (e.g., RTDETRDecoder, [nc, 500]).
Re-initialized and retrained the model with the updated configuration.
Lowered the confidence threshold (conf=0.01) during inference to retain more predictions.
Despite these adjustments, the output remains capped at 300 detections.

​Questions:

Is there an additional parameter or configuration step required to override the default num_queries=300 in the training script?
Are there potential conflicts between the decoder’s num_queries and other components (e.g., loss functions or matchers) that might enforce this limit?
Could you provide guidance on ensuring the model utilizes the updated num_queries value during both training and inference?
​Reproduction Context:

Model: RTDETR-l (custom dataset with nc=10).
Configuration: Modified num_queries=500 in the decoder block.
Training Command:
    from ultralytics.models import RTDETR
    model = RTDETR(model='./rtdetr_5000.yaml')
    # model.train(data='./data.yaml', epochs=1000, batch=1, device=0, imgsz=1920)
    model.train(
        data='./data.yaml',
        epochs=300,  # 高密度需更长训练
        imgsz=2560,  # 提升分辨率增强小目标检测
        batch=1,
        device=0
    )
# Ultralytics 🚀 AGPL-3.0 License - https://ultralytics.com/license

# Ultralytics RT-DETR-l hybrid object detection model with P3/8 - P5/32 outputs
# Model docs: https://docs.ultralytics.com/models/rtdetr
# Task docs: https://docs.ultralytics.com/tasks/detect

# Parameters
nc: 1  # <--- 你可以根据实际类别数改成你需要的，如 nc: 1
scales:  # model compound scaling constants
  l: [1.00, 1.00, 1024]

backbone:
  # [from, repeats, module, args]
  - [-1, 1, HGStem, [32, 48]]  # 0-P2/4
  - [-1, 6, HGBlock, [48, 128, 3]]  # stage 1

  - [-1, 1, DWConv, [128, 3, 2, 1, False]]  # 2-P3/8
  - [-1, 6, HGBlock, [96, 512, 3]]  # stage 2

  - [-1, 1, DWConv, [512, 3, 2, 1, False]]  # 4-P3/16
  - [-1, 6, HGBlock, [192, 1024, 5, True, False]]  # stage 3
  - [-1, 6, HGBlock, [192, 1024, 5, True, True]]
  - [-1, 6, HGBlock, [192, 1024, 5, True, True]]

  - [-1, 1, DWConv, [1024, 3, 2, 1, False]]  # 8-P4/32
  - [-1, 6, HGBlock, [384, 2048, 5, True, False]]  # stage 4

head:
  - [-1, 1, Conv, [256, 1, 1, None, 1, 1, False]]  # 10 input_proj.2
  - [-1, 1, AIFI, [1024, 8]]
  - [-1, 1, Conv, [256, 1, 1]]  # 12, Y5, lateral_convs.0

  - [-1, 1, nn.Upsample, [None, 2, "nearest"]]
  - [7, 1, Conv, [256, 1, 1, None, 1, 1, False]]  # 14 input_proj.1
  - [[-2, -1], 1, Concat, [1]]
  - [-1, 3, RepC3, [256]]  # 16, fpn_blocks.0
  - [-1, 1, Conv, [256, 1, 1]]  # 17, Y4, lateral_convs.1

  - [-1, 1, nn.Upsample, [None, 2, "nearest"]]
  - [3, 1, Conv, [256, 1, 1, None, 1, 1, False]]  # 19 input_proj.0
  - [[-2, -1], 1, Concat, [1]]
  - [-1, 3, RepC3, [256]]  # X3 (21), fpn_blocks.1

  - [-1, 1, Conv, [256, 3, 2]]  # 22, downsample_convs.0
  - [[-1, 17], 1, Concat, [1]]
  - [-1, 3, RepC3, [256]]  # F4 (24), pan_blocks.0

  - [-1, 1, Conv, [256, 3, 2]]  # 25, downsample_convs.1
  - [[-1, 12], 1, Concat, [1]]
  - [-1, 3, RepC3, [256]]  # F5 (27), pan_blocks.1

  - [[21, 24, 27], 1, RTDETRDecoder, [nc,5000]]  # ← 设置最多 5000 个候选目标

Any insights or code snippets to resolve this would be greatly appreciated. Thank you for your time and contributions to the community!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @HD-L, thank you for your detailed question and for using Ultralytics 🚀! This is an automated response to help get you started—an Ultralytics engineer will also review your issue and provide further assistance soon.

We recommend checking out the [Docs](https://docs.ultralytics.com/) for more information on YOLO and RT-DETR usage. Many common configuration and inference questions are addressed there, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) if you haven't already. This will help us diagnose the issue more effectively.

If you have a custom training ❓ Question like this one, please continue to provide as much information as possible, including:
- Model config files and exact changes made
- Training and inference commands
- Dataset examples (if possible)
- Any relevant logs or outputs

Also, please review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to ensure you’re following recommended practices.

Join the Ultralytics community for real-time help and idea sharing:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for quick chat
- [Discourse](https://community.ultralytics.com/) for in-depth discussions
- [Subreddit](https://reddit.com/r/Ultralytics) for community Q&A

## Upgrade

Make sure you are using the latest `ultralytics` package and that all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) are installed in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). This ensures your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in the following up-to-date verified environments (with all dependencies, including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your contribution and patience! An Ultralytics engineer will respond with further insights or next steps soon.

**Comment by Y-T-G** [comment]
You should also specify `max_det` during prediction

**Comment by HD-L** [comment]
from ultralytics.models import RTDETR
    model = RTDETR(model='./rtdetr_5000.yaml')
    model.train(
        data='./data.yaml',
        epochs=300,  # 高密度需更长训练
        imgsz=2560,  # 提升分辨率增强小目标检测
        batch=1,
        max_det=5000,
        device="cpu"
    )
is OK ？
